File: src/war3map/Generic/Rectangle.lua
local ____exports = {}
____exports.Rectangle = {}
local Rectangle = ____exports.Rectangle
Rectangle.name = "Rectangle"
Rectangle.__index = Rectangle
Rectangle.prototype = {}
Rectangle.prototype.__index = Rectangle.prototype
Rectangle.prototype.constructor = Rectangle
function Rectangle.new(...)
    local self = setmetatable({}, Rectangle.prototype)
    self:____constructor(...)
    return self
end
function Rectangle.prototype.____constructor(self, x, y, x2, y2)
    self.x = x
    self.y = y
    self.x2 = x2
    self.y2 = y2
end
function Rectangle.prototype.toRect(self)
    return Rect(self.x, self.y, self.x2, self.y2)
end
function Rectangle.fromRect(self, input)
    return ____exports.Rectangle.new(
        GetRectMinX(input),
        GetRectMinY(input),
        GetRectMaxX(input),
        GetRectMaxY(input)
    )
end
function Rectangle.fromRectClean(self, input)
    local retVar = ____exports.Rectangle.new(
        GetRectMinX(input),
        GetRectMinY(input),
        GetRectMaxX(input),
        GetRectMaxY(input)
    )
    RemoveRect(input)
    return retVar
end
return ____exports

File: war3map/Generic/Rectangle.ts
import {Delay} from "./Delay";

/**
 * My wrapper for Rects
 */
export class Rectangle {
    public x: number;
    public y: number;
    public x2: number;
    public y2: number;

    constructor(x: number, y: number, x2: number, y2: number) {
        this.x = x;
        this.y = y;
        this.x2 = x2;
        this.y2 = y2;
    }

    public toRectClean() {
        let rect = Rect(this.x, this.y, this.x2, this.y2);
        Delay.getInstance().addDelay(() => {
            RemoveRect(rect);
        }, 0.1);
        return rect;
    }

    public toRect() {
        return Rect(this.x, this.y, this.x2, this.y2);
    }

    public static fromRect(input: rect) {
        return new Rectangle(GetRectMinX(input), GetRectMinY(input), GetRectMaxX(input), GetRectMaxY(input));
    }

    public static fromRectClean(input: rect) {
        let retVar = new Rectangle(GetRectMinX(input), GetRectMinY(input), GetRectMaxX(input), GetRectMaxY(input));
        RemoveRect(input);
        return retVar;
    }
}